parser/impl/parser.cpp

```diff
@@ -469,17 +469,163 @@ sPtr<ast::IfStatement> Parser::parseIfStatement() {
+    ast::UnaryExpression expr;
+    if (opPrec(tok.type) >= 0) {
+        // if has operations on primary
+        if (tok.type == TokenType::Not || tok.type == TokenType::Sub ||
+            tok.type == TokenType::Add) {
+            // math unary operations
+            tok = m_lexer.Next();
+            expr.operation = tok;
+            expr.operand = parseUnaryExpression();
+            return std::make_shared<ast::UnaryExpression>(expr);
+
+        } else if (tok.type == TokenType::OpenParen) {
+            // parenthesis -> more priority
+            tok = m_lexer.Next();
+            sPtr<ast::Expression> expr = parseBinaryExpression(0);
+            if (m_lexer.Peek().type != TokenType::CloseParen) {
+                m_errors.push_back(Error{
+                    .pos = tok.pos,
+                    .message = "Expected to find ')'",
+                });
+                while (m_lexer.Peek().type != TokenType::CloseParen)
+                    m_lexer.Next();
+                return nullptr;
+            }
+            tok = m_lexer.Next(); // read ')'
+            return expr;
+        } else {
+            // throw error
+            m_errors.push_back(Error{
+                .pos = tok.pos,
+                .message = "Expected to find unary operator",
+            });
+            return nullptr;
+        }
+    } else if (isPrimitive(tok.type)) {
+        // if is pure primitive
+        tok = m_lexer.Next();
+        // check if More Then Identifier
+        if (tok.type == TokenType::Identifier) {
+            Token t = m_lexer.Peek();
+            switch (t.type) {
+            case TokenType::OpenParen:
+                return parseRoutineCall(tok);
+            case TokenType::Dot:
+            case TokenType::OpenBrack:
+                return parseModifiablePrimary(tok);
+            default:
+                break;
+            }
+        }
+        // return Primitive only
+        ast::Primitive prim;
+        prim.value = tok;
+        return std::make_shared<ast::Primitive>(prim);
+    }
     return nullptr;
 }
 
+sPtr<ast::Expression> Parser::parseBinaryExpression(int prec1) {
+
+    sPtr<ast::Expression> lhs = parseUnaryExpression();
+
+    for (;;) {
+        Token op = m_lexer.Peek();
+        int prec = opPrec(op.type);
+
+        if (prec < prec1) {
+            return lhs;
+        }
+
+        op = m_lexer.Next();
+
+        ast::BinaryExpression expr;
+        expr.operand1 = lhs;
+        expr.operation = op;
+        expr.operand2 = parseBinaryExpression(prec + 1);
+        lhs = std::make_shared<ast::BinaryExpression>(expr);
+    }
+}
+
+sPtr<ast::RoutineCall> Parser::parseRoutineCall(Token routineName) {
+    ast::RoutineCall rc;
+    rc.routine = routineName; // save the function name
+    Token t;
+    do {
+        t = m_lexer.Next(); // first call-read '(', others - ','
+        rc.args.push_back(parseExpression());
+
+    } while (m_lexer.Peek().type == TokenType::Comma);
```

Comment:
The next token may not be an opening parenthesis. You should check this before the loop. Additionally, do-while loops are highly discouraged because they're rare and allow more bugs to slip through on average. A general guideline: if you write do-while loops, then something is wrong and the code should be refactored.

Reply:
Why cannot it be if it is and it works?

Reply:
Oh. Got the idea